File: taijobi-web/src/lib/speak.ts
/**
 * Text-to-speech using the browser's SpeechSynthesis API.
 *
 * Language detection is delegated to the Zig `hanzi_detect_language`
 * export so TS + Swift + lexicon insert all agree on which voice to
 * pick. A tiny inline fallback covers the window where the older WASM
 * bundle hasn't been re-fetched from cache yet.
 */

import { detectLanguage } from './wasm';

const LANG_MAP: Record<string, string> = {
	zh: 'zh-CN',
	de: 'de-DE',
	en: 'en-US',
	ar: 'ar-SA',
	ja: 'ja-JP',
	ko: 'ko-KR',
	fr: 'fr-FR',
	es: 'es-ES',
	ru: 'ru-RU'
};

/** Single source of truth lives in libtaijobi/src/lang.zig. Falls back
 *  to a coarse heuristic only when the WASM export isn't loaded yet. */
export function detectSpeakLang(text: string): string {
	const zigGuess = detectLanguage(text);
	if (zigGuess) return zigGuess;
	const t = text.trim();
	if (!t) return 'en';
	if (/[一-鿿㐀-䶿]/.test(t)) return 'zh';
	if (/[؀-ۿ]/.test(t)) return 'ar';
	if (/[äöüÄÖÜß]/.test(t)) return 'de';
	return 'en';
}

export function speak(text: string, language: string | 'auto' = 'auto'): void {
	if (!('speechSynthesis' in window)) return;

	const lang = language === 'auto' ? detectSpeakLang(text) : language;

	window.speechSynthesis.cancel();

	const utterance = new SpeechSynthesisUtterance(text);
	utterance.lang = LANG_MAP[lang] ?? 'en-US';
	utterance.rate = 0.85;

	const voices = window.speechSynthesis.getVoices();
	const match = voices.find((v) => v.lang.startsWith(utterance.lang.split('-')[0]));
	if (match) utterance.voice = match;

	window.speechSynthesis.speak(utterance);
}

export function canSpeak(): boolean {
	return 'speechSynthesis' in window;
}

File: taijobi-ios/taijobi/Speak.swift
import AVFoundation

/// Tiny AVSpeechSynthesizer wrapper that mirrors taijobi-web/src/lib/speak.ts.
/// Maps a language code (zh/de/en/ar) to a BCP-47 locale, falls back to the
/// `detect()` heuristic when the caller passes "auto".
///
/// Held as a singleton so the synthesizer lives across taps — instantiating
/// one per tap leaks an utterance queue and sometimes drops the first call.
enum Speak {
    /// Single shared synthesizer — AVSpeechSynthesizer holds an underlying
    /// audio session and shouldn't be re-instantiated per tap.
    static let shared = AVSpeechSynthesizer()

    private static let langMap: [String: String] = [
        "zh": "zh-CN",
        "de": "de-DE",
        "en": "en-US",
        "ar": "ar-SA",
        "ja": "ja-JP",
        "ko": "ko-KR",
        "fr": "fr-FR",
        "es": "es-ES",
        "ru": "ru-RU",
    ]

    /// Language detection lives in libtaijobi (`lang.zig`) — one source
    /// of truth shared with TS + the lexicon insert path. The tiny
    /// fallback only runs when the library isn't initialised yet (no
    /// realistic code path hits it, but keeps `detect` total).
    static func detect(_ text: String) -> String {
        if let z = LibTaijobi.shared.detectLanguage(text) { return z }
        if text.range(of: #"\p{Han}"#, options: .regularExpression) != nil { return "zh" }
        if text.range(of: #"\p{Arabic}"#, options: .regularExpression) != nil { return "ar" }
        if text.range(of: "[äöüÄÖÜß]", options: .regularExpression) != nil { return "de" }
        return "en"
    }

    /// Speak the given text. Pass `"auto"` (or omit) to use the heuristic.
    static func say(_ text: String, language: String = "auto") {
        guard !text.isEmpty else { return }
        let lang = language == "auto" ? detect(text) : language
        let bcp = langMap[lang] ?? "en-US"

        // Stop the previous utterance — the user tapping a new word means
        // they're done listening to the old one.
        if shared.isSpeaking {
            shared.stopSpeaking(at: .immediate)
        }

        let utterance = AVSpeechUtterance(string: text)
        utterance.voice = AVSpeechSynthesisVoice(language: bcp)
        utterance.rate = AVSpeechUtteranceDefaultSpeechRate * 0.85
        shared.speak(utterance)
    }

    static var canSpeak: Bool { true }
}
